src/Microsoft.Extensions.ObjectPool/DefaultObjectPool.cs

```diff
@@ -18,43 +22,82 @@ public DefaultObjectPool(IPooledObjectPolicy<T> policy)
 
         public DefaultObjectPool(IPooledObjectPolicy<T> policy, int maximumRetained)
         {
-            if (policy == null)
+            _policy = policy ?? throw new ArgumentNullException(nameof(policy));
+            _isDefaultPolicy = IsDefaultPolicy();
+
+            // -1 due to _firstItem
+            _items = new ObjectWrapper[maximumRetained - 1];
+
+            bool IsDefaultPolicy()
             {
-                throw new ArgumentNullException(nameof(policy));
-            }
+                var type = policy.GetType();
 
-            _policy = policy;
-            _items = new T[maximumRetained];
+                return type.IsGenericType
+                    ? type.GetGenericTypeDefinition() == typeof(DefaultPooledObjectPolicy<>)
+                    : false;
+            }
         }
 
         public override T Get()
         {
-            for (var i = 0; i < _items.Length; i++)
+            T item = _firstItem;
+
+            if (item == null || Interlocked.CompareExchange(ref _firstItem, null, item) != item)
             {
-                var item = _items[i];
-                if (item != null && Interlocked.CompareExchange(ref _items[i], null, item) == item)
+                item = this.GetViaScan();
+            }
+
+            return item;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private T GetViaScan()
+        {
+            ObjectWrapper[] items = _items;
+            T item = null;
+
+            for (int i = 0; i < items.Length; i++)
+            {
+                item = items[i];
+
+                if (item != null && Interlocked.CompareExchange(ref items[i].Element, null, item) == item)
                 {
-                    return item;
+                    break;
                 }
             }
 
-            return _policy.Create();
+            return item ?? _policy.Create();
         }
 
         public override void Return(T obj)
         {
-            if (!_policy.Return(obj))
+            if (_isDefaultPolicy || _policy.Return(obj))
             {
-                return;
+                if (_firstItem != null || Interlocked.CompareExchange(ref _firstItem, obj, null) != null)
```

Comment:
The `Interlocked` is not absolutely necessary, but it is added due to https://github.com/aspnet/Common/issues/273

Without this `Interlocked` perf is even better.

Reply:
Yeah, that's right. The original design didn't have it for this reason and then we found that a lot of the use cases people cared about didn't work well when we 'forget' items, so we added the guarantee